File: creational/prototype.go
package main

import "fmt"

// Prototype interface
type Prototype interface {
	Clone() Prototype
	Show()
}

// User struct
type User struct {
	Name        string
	Role        string
	Permissions []string
}

func (u *User) Clone() Prototype {
	// ایجاد کپی عمیق از Slice
	newPermissions := make([]string, len(u.Permissions))
	copy(newPermissions, u.Permissions)

	return &User{
		Name:        u.Name,
		Role:        u.Role,
		Permissions: newPermissions,
	}
}

func (u *User) Show() {
	fmt.Printf("👤 %s (%s)\n", u.Name, u.Role)
	fmt.Printf("🔑 Permissions: %v\n\n", u.Permissions)
}

func main() {
	admin := &User{"Admin", "Administrator", []string{"create", "edit", "delete"}}
	admin.Show()

	user1 := admin.Clone().(*User)
	user1.Name = "Daniyal"
	user1.Role = "Editor"
	user1.Permissions = append(user1.Permissions, "publish")

	user1.Show()
	admin.Show() // admin هنوز بدون تغییره
}

File: creational/prototype.php
<?php

// Prototype interface
interface Prototype
{
    public function clone(): Prototype;
}

// کلاس اصلی
class User implements Prototype
{
    public string $name;
    public string $role;
    public array $permissions;

    public function __construct(string $name, string $role, array $permissions)
    {
        $this->name = $name;
        $this->role = $role;
        $this->permissions = $permissions;
    }

    public function clone(): Prototype
    {
        // Clone عمیق (deep copy)
        return clone $this;
    }

    public function show()
    {
        echo "👤 {$this->name} ({$this->role})\n";
        echo "🔑 Permissions: " . implode(', ', $this->permissions) . "\n\n";
    }
}

// نمونه‌ی اولیه
$admin = new User("Admin", "Administrator", ["create", "edit", "delete"]);
$admin->show();

// ساخت کاربر جدید با کپی
$user1 = $admin->clone();
$user1->name = "Daniyal";
$user1->role = "Editor";
$user1->permissions[] = "publish";

$user1->show();
$admin->show(); // هنوز دست‌نخورده‌ست


//👤 Admin (Administrator)
//🔑 Permissions: create, edit, delete

//👤 Daniyal (Editor)
//🔑 Permissions: create, edit, delete, publish

//👤 Admin (Administrator)
//🔑 Permissions: create, edit, delete
